### PR TITLE
OCPQE-16156 Wait mc to be applied on all nodes when there are multiple mcs at the…

### PR DESCRIPTION
**Pull request by JianLi-RH**
/cc @jiajliu @rioliu-rh

### CODE REVIEW

**Comment by JianLi-RH** [comment]
pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-4.14-upgrade-from-stable-4.13-gcp-ipi-xpn-fips-f28

**Comment by JianLi-RH** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-4.14-upgrade-from-stable-4.13-gcp-ipi-xpn-fips-f28

**Comment by JianLi-RH** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-4.14-upgrade-from-devpreview-4.14-gcp-ipi-workload-identity-rt-fips-f14

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 1]
```diff
@@ -338,19 +338,30 @@ function check_latest_machineconfig_applied() {
...
-    local max_retries; max_retries=$(expr $num \* 20)
-    while (( try < max_retries )); do
-        echo "Checking #${try}"
-        if ! check_latest_machineconfig_applied "${role}"; then
-            sleep ${interval}
-        else
-            break
+    local role="${1}" try=0 interval=30
+    num=$(oc get node --no-headers -l node-role.kubernetes.io/"$role"= | wc -l) 
+    local max_retries; max_retries=$(expr $num \* 20 \*60 \/ $interval) # Wait 30 minutes for each node, try 60/interval times per minutes
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:343]: 30/20/s

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 2]
```diff
@@ -338,19 +338,30 @@ function check_latest_machineconfig_applied() {
...
+
+    local mcp_try=0 mcp_status=''
+    local max_try_between_updated_and_updating; max_try_between_updated_and_updating=$(expr 5 \*60 \/ $interval) # We consider mcp to be updated if its status is updated for 5 minutes
+    while [ $mcp_try -lt $max_try_between_updated_and_updating ] && [ $try -lt $max_retries ]
+    do
+        sleep ${interval}
+        echo "Checking MCP #${try}"
+        mcp_status=$(oc get mcp/$mcp_status -o json | jq -r '.status.conditions[] | select(.type == "Updated") | .status')
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:351]: $mcp_status/$role/s

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh (hunk 1 above)
Comment (JianLi-RH) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:343]: what's wrong here?

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh (hunk 2 above)
Comment (JianLi-RH) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh]: fixed

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 3]
```diff
@@ -338,19 +338,30 @@ function check_latest_machineconfig_applied() {
...
+    local max_retries; max_retries=$(expr $num \* 20 \*60 \/ $interval) # Wait 30 minutes for each node, try 60/interval times per minutes
+
+    local mcp_try=0 mcp_status=''
+    local max_try_between_updated_and_updating; max_try_between_updated_and_updating=$(expr 5 \*60 \/ $interval) # We consider mcp to be updated if its status is updated for 5 minutes
+    while [ $mcp_try -lt $max_try_between_updated_and_updating ] && [ $try -lt $max_retries ]
+    do
+        sleep ${interval}
+        echo "Checking MCP #${try}"
+        mcp_status=$(oc get mcp/$role -o json | jq -r '.status.conditions[] | select(.type == "Updated") | .status')
+        echo "MCP updated is ${mcp_status}"
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:352]: how about move it out of the loop to decrease the log info? it should be ok to output the $role mcp's status once after the loop for debug.

Comment (JianLi-RH) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:352]: sure

**Comment by JianLi-RH** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-4.14-upgrade-from-devpreview-4.14-gcp-ipi-workload-identity-rt-fips-f14

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 4]
```diff
@@ -338,19 +338,30 @@ function check_latest_machineconfig_applied() {
...
+        echo "Checking MCP #${try}"
+        mcp_status=$(oc get mcp/$role -o json | jq -r '.status.conditions[] | select(.type == "Updated") | .status')
+        if [[ X"$mcp_status" != X"True" ]]; then
+            mcp_try=0
         fi
+        (( mcp_try += 1 ))
         (( try += 1 ))
     done
-    if (( try == max_retries )); then
+    echo "MCP updated is ${mcp_status}"
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:358]:
```suggestion
    echo "MCP ${role} status is ${mcp_status}"
```

**Comment by JianLi-RH** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-4.14-upgrade-from-devpreview-4.14-gcp-ipi-workload-identity-rt-fips-f14

**Comment by jiajliu** [comment]
Generally lgtm now. After the test pass, we should update step-registry/cucushift/chainupgrade/toimage/cucushift-chainupgrade-toimage-commands.sh accordingly, which also may hit the edge case due to upgrade.

**Comment by JianLi-RH** [comment]
@jiajliu @shellyyang1989 @jianlinliu please help review and merge, thanks

**Comment by shellyyang1989** [comment]
Thanks for the fix. We need to sync it to cucushift-chainupgrade-toimage-commands.sh.

**Comment by JianLi-RH** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-4.14-upgrade-from-ec-4.14-aws-ipi-f14

**Comment by JianLi-RH** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-4.14-upgrade-from-ec-4.14-aws-ipi-f14

**Comment by JianLi-RH** [comment]
@jiajliu upgrade-to-image passed
